util/unifyfs/src/unifyfs-rm.c

```diff
@@ -244,6 +256,111 @@ static int wait_server_initialization(unifyfs_resource_t* resource,
     return ret;
 }
 
+enum {
+    UNIFYFS_STAGE_IN = 0,
+    UNIFYFS_STAGE_OUT = 1,
+};
+
+static inline unsigned int estimate_timeout(const char* manifest_file, int op)
+{
+    /* FIXME: we just wait for 20 mins.
+     * In fact, we can roughly estimate time for stage in by looking at the
+     * total file size. However, such a method is not possible for stage out
+     * because we have no idea on what the total transfer size is.
+     */
+    return 20 * 60;
+}
+
+
+/**
+ * @brief wait until data stage operation finishes
+ *
+ * @param resource
+ * @param args
+ *
+ * @return
+ */
+static
+int wait_stage(unifyfs_resource_t* resource, unifyfs_args_t* args, int op)
+{
+    int ret = UNIFYFS_SUCCESS;
+    unsigned int interval = 5;
+    unsigned int wait_time = 0;
+    unsigned int timeout = 0;
+    FILE* fp = NULL;
+    const char* manifest_file = NULL;
+    char filename[PATH_MAX];
+    char linebuf[16];
+    int return_val_from_scnprintf;
+
+    return_val_from_scnprintf =
+        scnprintf(filename, PATH_MAX,
+                  "%s/%s", args->share_dir, UNIFYFS_STAGE_STATUS_FILENAME);
+    if (return_val_from_scnprintf > (PATH_MAX - 2)) {
+        fprintf(stderr, "Unifyfs status filename is too long!\n");
+        return -ENOMEM;
+    }
+
+
+    if (op == UNIFYFS_STAGE_IN) {
+        manifest_file = args->stage_in;
+    } else {
+        manifest_file = args->stage_out;
+    }
+
+    if (args->stage_timeout > 0) {
+        timeout = args->stage_timeout;
+    } else {
+        timeout = estimate_timeout(manifest_file, op);
+    }
+
+    while (1) {
+        fp = fopen(filename, "r");
+        if (fp) {
+            char* line = fgets(linebuf, 15, fp);
+            if (0 == strncmp("success", line, strlen("success"))) {
+                ret = 0;
+                break;      // transfer completed
+            } else if (0 == strncmp("fail", line, strlen("fail"))) {
+                ret = -EIO;
+                break;      // transfer failed
```

Comment:
Do you need to `fclose(fp)` for these cases?

Reply:
Correct, we need to.

Reply:
Fixed in upcoming commit.